### ISSUE TITLE
Windows emoji support

### ISSUE THREAD

**Issue by chrisduerr** [jwilm/alacritty#3082]
This is a follow-up to https://github.com/jwilm/alacritty/issues/153.

After Emojis are now implemented on macOS, Linux and BSD, the only platform lacking Emoji support is Windows.

**Comment by kchibisov** [comment]
This issue should be pretty simple btw. So if anyone would like to try implementing it, I could provide with all the required information, at least, I think that its all you should need.

P.s. I just dont have Windows machine to try implementing it myself.

**Comment by davidhewitt** [comment]
@kchibisov if you want to drop the information you have on this ticket, I could have a go at picking this up sometime.

**Comment by kchibisov** [comment]
The things I have for Windows emoji.

Everything should be changed only in the following [file](https://github.com/jwilm/alacritty/blob/master/font/src/directwrite/mod.rs) (folder). For colored emojis you should send `RGBA` buffer to Alacritty here, so you should use `RGB` buffer for all normal glyphs, and for colored once you should use `RGBA` [here](https://gist.github.com/makotokato/9407295).

This code that I found could be [useful](https://gist.github.com/makotokato/9407295) , but I'm not sure about it. Also the official [docs](https://docs.microsoft.com/en-us/windows/win32/directwrite/color-fonts).

If you found out that you should downsample things yourself we already have an algorithm around in https://github.com/jwilm/alacritty/blob/8a833406614b503dcd647015d3cfd239faf5cb0c/font/src/ft/mod.rs#L588.

You shouldn't change any code outside of the font crate in the end.

Also you should be aware of this bug #1864, so it's fine that your emojis would also suffer from this issue, since it's a bug in our renderer. I mean with `#ffffff` foreground and `#000000` background color the color of your emojis should be correct, if it's not correct with the provided colors then it's likely your PR fault. 

If you have more questions ask me(us) here or on IRC.

**Comment by davidhewitt** [comment]
Now that #3127 is merged (:tada:) we have greyscale emoji rendering!

Next is to add color rendering using the notes above from kchibisov. I had originally hoped to do the whole thing in one PR but chose to land PR #3127 on its own for two reasons:

- That in itself was a complicated enough change to warrant enough review and feedback for one PR
- I couldn't find the suggested APIs for color rendering in drwite (and maybe not even in winapi) - so some definitions might want to be added upstream for the color rendering to be added.

My time is split too many ways right now for me to champion a large-ish story like the above, though I would hope to return to it in the future.

That is very likely to be several months away though, so if someone wants to pick this up before I do, please go ahead, I can help brainstorm what the solution could be.

**Comment by LPGhatguy** [comment]
Hi, I'm using the latest master branch of Alacritty on Windows, and it does not appear that greyscale emoji render correctly for me. I'm using the Hack font.

I originally noticed this trying to adopt the [Starship prompt](https://starship.rs/) and filed this issue: https://github.com/starship/starship/issues/1514

From the status line, you can see that emoji are sometimes repeated, sometimes have strange glyphs attached, and have colors in the wrong places. It seems like Alacritty might be using byte indices instead of the resulting columns?

![alacritty_3pCoBxC3SO](https://user-images.githubusercontent.com/654598/88492161-9829b600-cf5d-11ea-967d-054a016cdda2.png)

As an isolated example, I made a small emoji test file that I tried to `cat` a file with these contents:

```
😅😅😅😅😅😅
🙂🙂🙂🙂🙂🙂
```

![alacritty_iMvcxKXWsR](https://user-images.githubusercontent.com/654598/88492301-ae844180-cf5e-11ea-93c6-7bbec2811294.png)

Is there a guide for how to configure Alacritty on Windows to work with greyscale emoji?

**Comment by davidhewitt** [comment]
Strange, no configuration is expected. What Windows 10 version are you running?

**Comment by chrisduerr** [comment]
This looks like a PTY issue to me tbh. But please open a separate issue since it has nothing to do with this (which is to track **colored** emojis).

**Comment by kchibisov** [comment]
>This looks like a PTY issue to me tbh

But if it's a pty issue it has nothing to do with alacritty, since we can't do anything about it.

**Comment by LPGhatguy** [comment]
I was running Windows 10 v1903. I just updated to v2004 and everything looks as intended. That makes me think that it was probably PTY issue as you guessed.

Maybe someone else will stumble on the same emoji rendering issues I did.

Thanks for the help, and sorry!

**Comment by klemola** [comment]
@chrisduerr @LPGhatguy I would like to know what is needed in order to have a proper fallback for emojis in WSL 2. I have read this issue repeatedly, but I am failing to understand why I am not seeing emoji output.

I am using Fira Code, font fallback works for Powerline symbols, yet emojis don't show up. I am launching Alacritty from Windows with WSL 2 as the shell. I have the latest Windows update, newer than v2004.

Thanks in advance!

**Comment by davidhewitt** [comment]
Does it work if you use the default font configuration? I haven't tried fira code; default configuration still rendering greyscale emojis fine for me.

**Comment by SuperSandro2000** [comment]
> Does it work if you use the default font configuration? I haven't tried fira code; default configuration still rendering greyscale emojis fine for me.

No, it does not use the emojis from the font. Even when I switch to an emoji only font it is still displaying the fallback greyscale pictures instead of the colored ones.

**Comment by davidhewitt** [comment]
Colored emojis are not implemented, only grayscale - this ticket is to track the implementation of colored emojis!

**Comment by Milo123459** [comment]
How could we go about integrating this? I'm happy to make a PR

**Comment by chrisduerr** [comment]
https://github.com/alacritty/alacritty/issues/3082#issuecomment-568049273

**Comment by meandmymind** [comment]
I've updated from 1909 to 20h2 and it works properly! 🎆

**Comment by Milo123459** [comment]
> I've updated from 1909 to 20h2 and it works properly! 🎆

I'm on 20h2 and it doesn't work :(

**Comment by sshock** [comment]
I was looking at https://github.com/alacritty/alacritty/issues/3082#issuecomment-568049273, but the code looks very different now.

It looks like alacritty is using `crossfont` now, so maybe adding support for colored emojis on Windows needs to happen in there?

**Comment by kchibisov** [comment]
It should, but the code should be the same though. Alacritty was using crossfont from the start, it's just crossfont moved to its own repository.

**Comment by sshock** [comment]
I found a more detailed example of using DirectWrite's TranslateColorGlyphRun [here](https://github.com/Microsoft/Windows-universal-samples/blob/main/Samples/DWriteColorGlyph/cpp/CustomTextRenderer.cpp#L49) .

But it is pretty involved and ends up relying on the higher-level Direct2D in order to support all the potential formats.

If we can't or don't want to use Direct2D for some reason, it looks like we would only be able to support the `COLR` format.

If we're ok with and able to use Direct2D, it sounds like we can avoid all this extra logic by just using DrawText.

**Comment by kchibisov** [comment]
Are you talking about OT-in-SVG as extra format, can't we use rust infra for that to render stuff, like on Linux? Supporting COLR is preferred though.

**Comment by sshock** [comment]
> Are you talking about OT-in-SVG as extra format

Yeah, OpenType-SVG.

> can't we use rust infra for that to render stuff, like on Linux?

I guess so, assuming we can pull out the SVG for the glyph.

> Supporting COLR is preferred though.

Good to know.  And we could just add support for COLR first and save SVG and CBDT/CBLC for later.

In order to continue with DirectWrite, we have to figure out how much of the logic ought to go into `dwrote-rs` vs in `crossfont`.  And the other problem is in order to use TranslateColorGlyphRun we need to be have an `IDWriteFactory4`, but `dwrote-rs` is still using just `IDWriteFactory`.

**Comment by kchibisov** [comment]
Well, I'd be surprised that you can't get any of the color formats with alpha, because emojis should have alpha in their mask. Because it's basically a picture with transparent regions.

Just doing `COLR` is fine, we don't have `ot-in-svg` anyway.

Do you happen to know what microsoft is using by default for emojis, is it `COLR` table?

**Comment by sshock** [comment]
> Well, I'd be surprised that you can't get any of the color formats with alpha, because emojis should have alpha in their mask. Because it's basically a picture with transparent regions.

Yeah, I don't doubt that all the info is there in the font, but I just don't see how to get it out with DirectWrite.

All I see so far is the [CreateAlphaTexture](https://learn.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwriteglyphrunanalysis-createalphatexture), which only allows you to get the data from a glyph run out as mono or RGB.

Most of the examples with glyph runs use Direct2D Draw methods in the end to then draw the glyph run to a device context.  I honestly don't know, but perhaps that strategy is more flexible and with the right type of device context (like an in-memory DIB) it might be possible to get the data out as RGBA.

> Just doing `COLR` is fine, we don't have `ot-in-svg` anyway.

Supporting just COLR for now sounds good.

What do you mean exactly by "we don't have 'ot-in-svg'"?  By that do you mean that even on non-windows platforms right now alacritty doesn't support OpenType font whose color glyphs embedded as SVGs? 

> Do you happen to know what microsoft is using by default for emojis, is it `COLR` table?

I think so, but let me do some research or testing and I'll try to find out.  All I know so far is that Windows uses the "Segoe UI Emoji" font.

**Comment by sshock** [comment]
> > Do you happen to know what microsoft is using by default for emojis, is it `COLR` table?
> 
> I think so, but let me do some research or testing and I'll try to find out. All I know so far is that Windows uses the "Segoe UI Emoji" font.

Ok, yep I was able to use `otfinfo` to confirm it is using COLR tabl:

```
sshock@sshock:/share/tmp$ otfinfo -t seguiemj.ttf
 157334 COLR
   2256 CPAL
   9636 DSIG
    168 GDEF
   2952 GPOS
  32490 GSUB
     96 OS/2
  21298 cmap
    490 cvt
   2435 fpgm
     12 gasp
2480636 glyf
     54 head
     36 hhea
  58008 hmtx
   8382 kern
  58012 loca
     32 maxp
     66 meta
   2799 name
     32 post
   1279 prep
```

**Comment by kchibisov** [comment]
>What do you mean exactly by "we don't have 'ot-in-svg'"? By that do you mean that even on non-windows platforms right now alacritty doesn't support OpenType font whose color glyphs embedded as SVGs?

Yes, we don't support OT-in-SVG at all on other platforms, because 1. On linux it's a new format added like a month ago before, which requires providing your own svg renderer. 2. On macOS this format is likely internally supported, but no changes were required from alacritty, and I don't know what is happening on it. 

>Most of the examples with glyph runs use Direct2D Draw methods in the end to then draw the glyph run to a device context.

That's correct, you likely need to draw something into a temporary buffer, like we have with other backends, e.g. freetype. I don't know what is the situation with Windows though, so what I'm saying might be wrong.

**Comment by umanoinkei** [comment]
@kchibisov Can I sponsor a windows machine for you?

**Comment by kchibisov** [comment]
I do have windows machine, I'm just not that interested in doing anything on windows myself, since I have very little time to spend it on things I don't use myself (windows).
